### ISSUE TITLE
[feature request] SGE cluster execution

### ISSUE THREAD

**Issue by JingqiuLiao** [metagenome-atlas/atlas#327]
Hi,
I’m using the Sun Grid Engine (SGE) cluster system. May I know if you can provide an easy set up for this system? Thanks.

**Comment by SilasK** [comment]
I can create one. What is the command that you would use to submit to your cluster if you want to specify:
- cpus
- total virtual memory
- running time

something like 

`qsub -l h_vmem=100G -l s_rt=600`

**Comment by SilasK** [comment]
No, not yet. But  we can make one. 

Can you confirm me that when you submit a job you get something like:


When you ask for the jobstatus with `qacct -j {jobid}` you get a text that contains somewhere 
`

**Comment by JingqiuLiao** [comment]
Thank you very much! This is how we submit a job:
```bash
$ qsub -l mem=128M,time=01:15:00 job.sh    # <-- 128M of memory, 1 hr 15 min
```
After the job is submitted, I get:
```bash
Your job {jobid} ("job.sh") has been submitted
```
to get the jobstatus, yes, we use:
```bash
qacct –j {jobid}
```

**Comment by SilasK** [comment]
And how do you define processors?

**Comment by JingqiuLiao** [comment]
```bash
$ qsub -pe smp {number_of_processors}
```
Thank you!

**Comment by SilasK** [comment]
Looking on different forums I found that the command:

```
qsub -terse -V -cwd -q {queue} -N {name} -pe {threads} -l h_vmem={memory}G -l h_rt={time_minutes}00 jobscript.sh
```

should also work and only return the job id. 

Could you test if this works on your system? 

If this works, you can install the profile:
```
mkdir -p ~/.config/snakemake
cd ~/.config/snakemake
cookiecutter https://github.com/metagenome-atlas/clusterprofile.git --checkout sge
```

And the run atlas with the argument:
```
atlas run --profile cluster
```

**Comment by JingqiuLiao** [comment]
Thank you so much! This is very helpful! Our cluster uses SMP as the parallel environment, so it should be `-pe smp {threads}`. I’ve forked the repo and updated the key mapping. However, atlas reported the following error when I ran test samples using `atlas run qc --profile cluster`:

```
submit command: qsub -terse -V -cwd  -N initialize_qc -pe smp 4 -l h_vmem=10G -l h_rt=3000 /home/username/test/.snakemake/tmp.s0rx9848/snakejob.initialize_qc.24.sh
Traceback (most recent call last):
  File "/home/username/.config/snakemake/cluster/scheduler.py", line 66, in <module>
    raise Exception("Job can't be submitted\n"+output.decode("utf-8")+error.decode("utf-8"))
Exception: Job can't be submitted
Unable to read script file because of error: error opening : No such file or directory
```
Could you please help with this? Thanks.

**Comment by SilasK** [comment]
It seems as if the job script (/home/username/test/.snakemake/tmp.s0rx9848/snakejob.initialize_qc.24.sh)  cannot be found.

No Idea why this happens.

You could run atlas with the argument --skip-script-cleanup and after the error.

An then try to run the command
qsub -terse -V -cwd  -N initialize_qc -pe smp 4 -l h_vmem=10G -l h_rt=3000 /home/username/test/.snakemake/tmp.s0rx9848/snakejob.initialize_qc.24.sh

To see if you get more information

**Comment by JingqiuLiao** [comment]
Hi @SilasK, thank you very much for your help. I think I found the solution for this. It is because in the “key_mapping.yaml” for sge, there is an extra space at the end of _command: "qsub -terse -V -cwd “_. With that space removed, it is able to find the job script.

I was able to submit jobs and initialized QC. However, at the next step I ran into other two errors.

The first error is:
```bash
Error in rule get_read_stats:
    jobid: 15
    output: sample1/sequence_quality_control/read_stats/raw.zip, sample1/sequence_quality_control/read_stats/raw_read_counts.tsv
    log: sample1/logs/QC/read_stats/raw.log (check log file(s) for error message)
    cluster_jobid: 14891671
```
I looked into raw.log and it says “/bin/bash: line 2: reformat.sh: command not found”. I found in issue #272 that Sofie also ran into this. I checked reformat.sh and it is installed already in the conda environment which atlas created in the folder called “databases" in the working directory as well as in the atlas environment under anaconda. I’m not sure why when I submit my job to the cluster, it can’t find reformat.sh (note: I was submitting jobs from the head node). I also tried your suggestion and re-initialize conda using "conda activate base; conda init bash”. But still got the same error.

The second error is:
```bash
Error in rule deduplicate_reads:
    jobid: 29
    output: sample1/sequence_quality_control/sample1_deduplicated_R1.fastq.gz, sample1/sequence_quality_control/sample1_deduplicated_R2.fastq.gz
    log: sample1/logs/QC/deduplicate.log (check log file(s) for error message)
    conda-env: /storage/scratch/c2b2/korem/jl5897/MOMS-PI/test/databases/conda_envs/559ee9a1
    shell:

                clumpify.sh                     in1=sample1/sequence_quality_control/sample1_raw_R1.fastq.gz in2=sample1/sequence_quality_control/sample1_raw_R2.fastq.gz                     out1=sample1/sequence_quality_control/sample1_deduplicated_R1.fastq.gz out2=sample1/sequence_quality_control/sample1_deduplicated_R2.fastq.gz                     overwrite=true                    dedupe=t                     dupesubs=2                     optical=f                    threads=8                     -Xmx51G 2> sample1/logs/QC/deduplicate.log

        (one of the commands exited with non-zero exit code; note that snakemake uses bash strict mode!)
    cluster_jobid: 14891672

Error executing rule deduplicate_reads on cluster (jobid: 29, external: 14891672, jobscript: /storage/scratch/c2b2/korem/jl5897/MOMS-PI/test/.snakemake/tmp.xnyefbqd/snakejob.deduplicate_reads.29.sh). For error details see the cluster log and the log files of the involved rule(s).
```
I added “-S /bin/bash” to the qsub command but seem it did not help. 

I appreciate your time looking into the two errors. Thank you!

**Comment by SilasK** [comment]
Great you found the error. What all space can cause for problem. Do you mind to make a pull request four your updated cluster profile?


In the first error:

get_read_stats: uses in fact the reformat.sh that is installed in the atlas environment. In theory the argument -V supplied to qsub should load the environment to the cluster. As it doesn’t work, maybe you could drop it and see what happens.


For the second error:

Check the  sample1/logs/QC/read_stats/raw.log
And there should eb a cluster file log with the number 14891671
Do you find it ?

**Comment by JingqiuLiao** [comment]
I just sent a pull request. I think it is working now. Thank you very much for your help.

**Comment by SilasK** [comment]
Thank you very much for your help.

**Comment by yanhui09** [comment]
Hi @SilasK 
 
Thanks for the great workflow. I tried it on a big cluster, it works smoothly with several rules , but met a similar issue: 
"/usr/bin/bash: line 2: reformat.sh: command not found".
I can find this command when I activate the conda env at the login node, but it didn't work while submited to the computing node. Is there any solution for this?

The cluster use pbs-torque batch system with a “first come, first serve” policy (no need to reference it with "-q").
I tried to assign "-V" with "qsub", but it didn't work.

BTW,
I couldn't find any snakemake job script under .snakemake folder.  Is it normal? Where are the possible location for these files?
I tried the latest conda version 2.6a2 and dev version 2.6a4, and met the same issue in both versions.

**Comment by SilasK** [comment]
Usually, scripts are deleted after execution. there is a snakemake argument to not delete them for debug purposes.

When you log in is the conda base environment activated? 

It might be related to the fact that when a script is submitted to the cluster a new bash is activated where the conda env is not loaded.  Try to add the conda command to the `.bashrc` I think  `conda init` should do this.

**Comment by yanhui09** [comment]
Yes. When I log in, I'm in the conda base environment, e.g. (base) [yanhui@host]
I tried to initialize the conda env, following the suggestions, but seemed not to solve the problem. 

But I found a walk-round solution. While I compared the rule `get_read_stats` and `initialize_qc` (both using `reformat.sh`), the perfomance success is decided by whether conda environment is adopted.  It seems I couldn't let the computing node inherit the conda environment of my login node (probably due to the cluster set-up, admins strongly advised against `qsub -v`), but I can trigger a conda environment through snakemake rules. 

My solution is to use environmental module to first load bbmap , and run atlas in the cluster mode. Now it runs smoothly!!
I think you're right. It's due to the conda env initialization in the computing node. Also I can see, you probably will wrap the code block of "get_read_stats" into an external script, so that  this rule can also be run in the isolated conda environment.  Maybe it's not on the priority list for now. I will check if I will give a hand there.

Again many thanks for atlas! It will save me much time. 🍺
